### ISSUE TITLE
Getting exception for a redundant subjectAltName

### ISSUE THREAD

**Issue by erezsh** [pyca/pyopenssl#747]
This is my 3rd stop trying to get this fixed.

1. https://github.com/requests/requests/issues/4569

2. https://github.com/kjd/idna/issues/60

"Emojis are not valid in domain names... Really this is pyopenssl's fault, it should squash the error since it isn't relevant (invalid subjectAltName that doesn't match the domain we're fetching anyway)."

**Comment by alex** [comment]
Can you please provide instructions for us to reproduce and the traceback you're seeing?

**Comment by erezsh** [comment]
Hi Alex,
I don't know how to reproduce this directly in pyopenssl.

However, it happens with the following code using requests (when it has pyopenssl enabled):

    import requests
    requests.get('https://thegroovecartel.com/feed/')

**Comment by reaperhulk** [comment]
This is coming from https://github.com/urllib3/urllib3/blob/master/urllib3/contrib/pyopenssl.py#L157 which was originally put in place to deal with `cryptography`'s API mistake of returning unicode str (corrected in 2.1). Since `xn--mn8ha4uc` is indeed not a valid A-label the `idna.encode` call fails.

What we really want, however, is it to pass through the bytes. In cryptography 2.1+ calling `_dnsname_to_stdlib` is unnecessary so for users on that codepath you could simply remove that entire function. On older versions of cryptography this would actually have failed during the call to get the SAN extension (with an `InvalidCodepoint` exception), so I'm not sure what the `urllib3` maintainers might want to do to avoid that.

**Comment by erezsh** [comment]
Just one note: The exception is different on Python3

**Comment by fxcoudert** [comment]
With python3 I am getting this:

```
Python 3.6.5 |Anaconda custom (64-bit)| (default, Apr 26 2018, 08:42:37) 
[GCC 4.2.1 Compatible Clang 4.0.1 (tags/RELEASE_401/final)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import requests
>>> requests.get("https://earthlingsoft.net/UnicodeChecker/appcast.xml")
Traceback (most recent call last):
  File "/Users/fx/anaconda3/lib/python3.6/site-packages/idna/core.py", line 271, in alabel
    ulabel(label)
  File "/Users/fx/anaconda3/lib/python3.6/site-packages/idna/core.py", line 311, in ulabel
    check_label(label)
  File "/Users/fx/anaconda3/lib/python3.6/site-packages/idna/core.py", line 261, in check_label
    raise InvalidCodepoint('Codepoint {0} at position {1} of {2} not allowed'.format(_unot(cp_value), pos+1, repr(label)))
idna.core.InvalidCodepoint: Codepoint U+2603 at position 1 of '☃' not allowed

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/fx/anaconda3/lib/python3.6/site-packages/requests/api.py", line 72, in get
    return request('get', url, params=params, **kwargs)
  File "/Users/fx/anaconda3/lib/python3.6/site-packages/requests/api.py", line 58, in request
    return session.request(method=method, url=url, **kwargs)
  File "/Users/fx/anaconda3/lib/python3.6/site-packages/requests/sessions.py", line 512, in request
    resp = self.send(prep, **send_kwargs)
  File "/Users/fx/anaconda3/lib/python3.6/site-packages/requests/sessions.py", line 622, in send
    r = adapter.send(request, **kwargs)
  File "/Users/fx/anaconda3/lib/python3.6/site-packages/requests/adapters.py", line 445, in send
    timeout=timeout
  File "/Users/fx/anaconda3/lib/python3.6/site-packages/urllib3/connectionpool.py", line 600, in urlopen
    chunked=chunked)
  File "/Users/fx/anaconda3/lib/python3.6/site-packages/urllib3/connectionpool.py", line 343, in _make_request
    self._validate_conn(conn)
  File "/Users/fx/anaconda3/lib/python3.6/site-packages/urllib3/connectionpool.py", line 849, in _validate_conn
    conn.connect()
  File "/Users/fx/anaconda3/lib/python3.6/site-packages/urllib3/connection.py", line 367, in connect
    cert = self.sock.getpeercert()
  File "/Users/fx/anaconda3/lib/python3.6/site-packages/urllib3/contrib/pyopenssl.py", line 351, in getpeercert
    'subjectAltName': get_subj_alt_name(x509)
  File "/Users/fx/anaconda3/lib/python3.6/site-packages/urllib3/contrib/pyopenssl.py", line 228, in get_subj_alt_name
    for name in ext.get_values_for_type(x509.DNSName)
  File "/Users/fx/anaconda3/lib/python3.6/site-packages/urllib3/contrib/pyopenssl.py", line 228, in <listcomp>
    for name in ext.get_values_for_type(x509.DNSName)
  File "/Users/fx/anaconda3/lib/python3.6/site-packages/urllib3/contrib/pyopenssl.py", line 181, in _dnsname_to_stdlib
    name = idna_encode(name)
  File "/Users/fx/anaconda3/lib/python3.6/site-packages/urllib3/contrib/pyopenssl.py", line 179, in idna_encode
    return idna.encode(name)
  File "/Users/fx/anaconda3/lib/python3.6/site-packages/idna/core.py", line 361, in encode
    s = alabel(label)
  File "/Users/fx/anaconda3/lib/python3.6/site-packages/idna/core.py", line 273, in alabel
    raise IDNAError('The label {0} is not a valid A-label'.format(label))
idna.core.IDNAError: The label b'xn--n3h' is not a valid A-label
```

So two errors:
- `idna.core.InvalidCodepoint: Codepoint U+2603 at position 1 of '☃' not allowed`
- `idna.core.IDNAError: The label b'xn--n3h' is not a valid A-label`

**Comment by SethMichaelLarson** [comment]
This has been fixed in urllib3/urllib3#1406.
